### PR TITLE
Add a `str()` method

### PR DESCRIPTION
**Pull request by dlecocq**
@b4hand @tanglyh @tammybailey @martin-seomoz

### CODE REVIEW

File: include/url.h [hunk 1]
```diff
@@ -63,6 +63,11 @@ namespace Url
 
         const std::string& userinfo() const { return userinfo_; }
 
+        /**
+         * Get a new string representation of the URL.
+         **/
+        std::string toString() const;
```
Comment (b4hand) [inline include/url.h]: In C++, this method is often just called [`str()`](http://en.cppreference.com/mwiki/index.php?title=Special%3ASearch&search=str).

**Comment by b4hand** [comment]
LGTM, but I wonder if the trade off of recomputing `toString()` is a great one. The downside to keeping the fully formed URL is twice the space for the object.

File: include/url.h (hunk 1 above)
Comment (dlecocq) [inline include/url.h]: When in Rome...

I'll change that.

**Comment by dlecocq** [comment]
I'm not sure what you mean -- are you suggesting we memoize the string representation?

**Comment by b4hand** [comment]
Yeah, that's what I was considering whether it was beneficial to store the full URL potentially in the constructor. Then the question is whether you eagerly store it or only on demand.

**Comment by dlecocq** [comment]
It's going to change when people start doing normalizations, so I figure it's best to do so lazily. Another use-case is to extract the PLD / TLD / FQDN, in which case `str()` would never even be invoked.

**Pull request merged**